torch/csrc/distributed/rpc/message.cpp

```diff
@@ -68,10 +68,23 @@ void Message::setId(int64_t id) {
 
 std::vector<c10::weak_intrusive_ptr<c10::StorageImpl>> Message::getStorages()
     const {
+  // Sparse tensors do not have storage. Instead, a sparse tensor
+  // contains two tensors indices and values, and both contain storage.
   std::vector<c10::weak_intrusive_ptr<c10::StorageImpl>> storages;
-  storages.reserve(tensors_.size());
+  auto numStorages = tensors_.size();
   for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+    if (tensor.is_sparse()) {
+      ++numStorages;
+    }
+  }
+  storages.reserve(numStorages);
```

Comment:
We use reserve instead of emplace_back to prevent vector reallocation. If the vector capacity is smaller than the new size, then automatic reallocation will occur.

```
1
2
4
8
16
32
64
128
```

The minimum size is N when all tensors are dense. The maximum size is 2N when all tensors are sparse.  Reserve runs in linear time https://en.cppreference.com/w/cpp/container/vector/reserve. 

Both are identical in Big O notation time complexity. I analyzed without counting constant operations such as emplace_back.

Solution 1 two loops:
The worst case for solution 1 is Loop N elements.  Reserve max 2N. Loop N tensors emplace_back.
O(4N) = O(N)
The best case for solution 1 is Loop N elements. Reserve N elements. Loop N tensors emplace_back.
O(3N) = O(N)

Solution 2 one loop reserve 2N:
The worst case is reserve 2N. Then loop N tensors emplace_back.
O(3N) = O(N)

Both are identical in Big O space complexity.  

I was leaning towards solution 1 because of the additional memory needed to run the program, not the algorithm. Also, I have the first use case for sparse tensors going through this code path, so I assume most tensors will be dense.

Rethinking this though I will merge the loops based on time complexity.